client/trino-jdbc/src/test/java/io/trino/jdbc/TestTrinoDatabaseMetaData.java

```diff
@@ -1132,8 +1133,9 @@ public void testGetTablesMetadataCalls()
                         list(COUNTING_CATALOG, "test_schema1", "test_table1", "TABLE"),
                         list(COUNTING_CATALOG, "test_schema2", "test_table1", "TABLE")),
                 new MetadataCallsCount()
-                        .withListSchemasCount(1)
-                        .withListTablesCount(2));
+                        .withListSchemasCount(3)
+                        .withListTablesCount(6)
```

Comment:
can you please exemplify what were the 2 list tables calls before the change, and what are the 6 calls after the change?

Reply:
> applyFilter() call during sql plan at TableJdbcTable.java:176 with filter prefix "mock_catalog.test_schema1.*"

this is called twice (as you observed). what would it take to avoid doing that?

Reply:
Because the test generated query has sort in it, during the planning:
it would first go into https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AddExchanges.java#L478 and do first pass of visitFilter->pushFilterIntoTableScan->applyFilter;
seeing the returned child plan is not single node, then it would get to AddExchanges.java#L493 and do a second pass of applyFilter.
I'm not so sure about why the logic is what it is in AddExchanges.java. Feels like some optimizations could happen there.